sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2120,6 +2120,19 @@ class Dataset[T] private[sql](
   }
 
   /**
+   * Returns a new Dataset by adding columns with metadata.
+   */
+  private[spark] def withColumns(
+      colNames: Seq[String],
+      cols: Seq[Column],
+      metadata: Seq[Metadata]): DataFrame = {
+    val newCols = colNames.zip(cols).zip(metadata).map { case ((colName, col), metadata) =>
```

Comment:
Is that possible the number of elements in metadata do not match? Then, the results will be unexpected because of  this impl

Reply:
Yes. We should check the number of metadata too. I'll add it when adding related test.